code/controllers/subsystem/machines.dm

```diff
@@ -13,6 +20,42 @@ SUBSYSTEM_DEF(machines)
 	fire()
 	return SS_INIT_SUCCESS
 
+/datum/controller/subsystem/machines/proc/register_machine(obj/machinery/machine)
+	LAZYADD(machines_by_type[machine.type], machine)
+	all_machines |= machine
+
+/datum/controller/subsystem/machines/proc/unregister_machine(obj/machinery/machine)
+	var/list/existing = machines_by_type[machine.type]
+	existing -= machine
+	if(!length(existing))
+		machines_by_type -= machine.type
+	all_machines -= machine
+
+/datum/controller/subsystem/machines/proc/get_machines_by_type_and_subtypes(obj/machinery/machine_type)
+	if(!ispath(machine_type))
+		machine_type = machine_type.type
+	if(!ispath(machine_type, /obj/machinery))
+		CRASH("called get_machines_by_type_and_subtypes with a non-machine type [machine_type]")
+	var/list/machines = list()
+	var/list/subtypes = typesof(machine_type)
+	for(var/next_type in subtypes)
+		if(!(next_type in machines_by_type))
+			continue
+		machines += machines_by_type[next_type]
```

Comment:
```suggestion
		var/list/found_machines = machines_by_type[next_type]
		if(found_machines)
			machines += found_machines 
```
Guess we can check the list for the key once rather than twice. It adds up